### PR TITLE
Added the Admin transfer limit feature for Superadmins with edit history

### LINKED ISSUE #14894: [Governance::Admin::Transfer::Limit] Add UI for superadmins to configure transfer limits
> [!NOTE]
> This has been assigned to @mahimtiwari

Right now, there is no UI for superadmins to configure other admin's transfer limits. Gary has to go into the Rails console and run:

```ruby
user = # sldfjslkfd
amount_cents = 10_000_00 # $10k
Governance::Admin::Transfer::Limit.find_or_initalize_by(user:).update!(amount_cents:)
```

Let's add a UI form for this on the User's admin settings page. e.g. `/users/gary/settings/admin`

Ensure this setting can't be changed while impersonating

Users should NOT be able to edit their own transfer limit. Ensure paper trail logs these changes

### PR DESCRIPTION
**Pull request by mahimtiwari**
## Summary of the problem

Till now there was no UI for superadmins to set the "Admin Transfer Limit", so they had to manually change it in the rails console.

closes #14894 

## Describe your changes

A new section named "Admin Transfer Limit" was added to the Admin Settings page `/users/{slug}/admin` to let superadmins change the "Admin Transfer Limit" from the admin portal itself and also enable them to easliy view the edit history. It was also ensured that the changes were being logged by papertrails. It was also ensured that proper update flashes were visible to the superadmins.

image

Keeping in mind the criticallity of this settings to HCB, security measures were also added:-
- The API Endpoint that allows for this changes is only accesible/limited to superadmins.
- The field where the transfer limit can be changed is only rendered for Superadmins, though this is mainly for UI purposes.
- It was also further tested for any vulnerabilities, though none were found.

| Endpoint | Access |
|---|---|
| `update_admin_transfer_limit` | superadmin ONLY |
| `admin_transfer_limit_history` | admin+ superadmin |

| Admin | Superadmin |
|---|---|
|Paper Trails | Transfer Limit |

Aditionally, in the edit history of the admin transfer limit pagination support was also added to account for the case where there are multiple changes preventing the flooding of the admin page with those logs.

image

### CODE REVIEW

File: app/controllers/users_controller.rb [hunk 1]
```diff
@@ -368,6 +368,52 @@ def suppress_card_locking
...
+                 version_class
+                   .where(
+                     item_type: "Governance::Admin::Transfer::Limit",
+                     item_id: limit.id
+                   )
+                   .order(created_at: :desc)
+                   .page(params[:page])
+                   .per(params[:per] || 25)
+               else
+                 version_class
```
Comment (Luke-Oldenburg) [inline app/controllers/users_controller.rb]:
```suggestion
                 PaperTrail::Version
```

File: app/controllers/users_controller.rb [hunk 2]
```diff
@@ -368,6 +368,52 @@ def suppress_card_locking
...
+    )
+
+    version_class = Governance::Admin::Transfer::Limit.paper_trail.version_class
+
+    @history = if limit
+                 version_class
+                   .where(
+                     item_type: "Governance::Admin::Transfer::Limit",
+                     item_id: limit.id
+                   )
```
Comment (Luke-Oldenburg) [inline app/controllers/users_controller.rb]:
```suggestion
                 limit.versions
```

File: app/controllers/users_controller.rb [hunk 3]
```diff
@@ -368,6 +368,52 @@ def suppress_card_locking
...
+  def admin_transfer_limit_history
+    authorize @user
+
+    target_user_id = params[:target_user_id].to_i
+
+    limit = Governance::Admin::Transfer::Limit.find_by(
+      user_id: target_user_id
+    )
+
+    version_class = Governance::Admin::Transfer::Limit.paper_trail.version_class
```
Comment (Luke-Oldenburg) [inline app/controllers/users_controller.rb]:
```suggestion
```

File: app/views/governance/admin/transfer/limits/history.html.erb
Comment (Luke-Oldenburg) [inline app/views/governance/admin/transfer/limits/history.html.erb:1]: Make this a regular page.

File: app/views/users/_admin_transfer_limit_history.html.erb [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+          <th>Set By</th>
+        </tr>
+      </thead>
+
+      <tbody>
+        <% @history.each do |version| %>
+          <% changes = version.changeset %>
+          <% changer = @changers[version.whodunnit.to_s] %>
+
+          <tr class="">
```
Comment (Luke-Oldenburg) [inline app/views/users/_admin_transfer_limit_history.html.erb]:
```suggestion
          <tr>
```

File: app/views/users/_admin_transfer_limit_history.html.erb [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+              <div class="flex items-center gap-2">
+                <% if changer.present? %>
+                  <%= avatar_for changer, class: "align-middle author", cached: true %>
+                  <span><%= changer.name || changer.email %></span>
+                <% else %>
+                  <span>Unknown</span>
+                <% end %>
+              </div>
+            </td>
+
```
Comment (Luke-Oldenburg) [inline app/views/users/_admin_transfer_limit_history.html.erb]:
```suggestion
```

File: app/views/users/_admin_transfer_limit_history.html.erb [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+          </tr>
+        <% end %>
+      </tbody>
+    </table>
+  </section>
+
+  <%= paginate @history,
+        params: { target_user_id: @user.id } %>
```
Comment (Luke-Oldenburg) [inline app/views/users/_admin_transfer_limit_history.html.erb]:
```suggestion
  <%= paginate @history, params: { target_user_id: @user.id } %>
```

File: app/views/users/edit_admin.html.erb [hunk 7]
```diff
@@ -107,6 +107,35 @@
     </section>
   <% end %>
 
+  <% admin_tool "p-3" do %>
+    <h3 class="mb2 mt1">Admin Transfer Limit</h3>
+    <% limit = Governance::Admin::Transfer::Limit.find_by(user_id: @user.id) %>
```
Comment (Luke-Oldenburg) [inline app/views/users/edit_admin.html.erb]: This should be defined as `@limit` in `UsersController#edit_admin`

File: app/controllers/users_controller.rb
Comment (garyhtou) [inline app/controllers/users_controller.rb:1]: Can we make this its own separate controller? Ideally, a `Governance::Admin::Transfer::LimitsController`

Comment (mahimtiwari) [reply app/controllers/users_controller.rb:1]: done!

File: app/controllers/governance/admin/transfer/limits_controller.rb [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+                                          .page(params[:page])
+                                          .per(params[:per] || 25)
+                     end
+
+          changer_ids = @history.map(&:whodunnit).compact.uniq
+
+          @changers = User
+                      .where(id: changer_ids)
+                      .index_by { |user| user.id.to_s }
+
```
Comment (Luke-Oldenburg) [inline app/controllers/governance/admin/transfer/limits_controller.rb]:
```suggestion
```

File: app/controllers/governance/admin/transfer/limits_controller.rb [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+          @history = if @limit
+                       @limit.versions
+                             .reorder(created_at: :desc)
+                             .page(params[:page])
+                             .per(params[:per] || 25)
+                     else
+                       PaperTrail::Version.none
+                                          .page(params[:page])
+                                          .per(params[:per] || 25)
+                     end
```
Comment (Luke-Oldenburg) [inline app/controllers/governance/admin/transfer/limits_controller.rb]: `@limit` will always exist because of how `set_limit` functions. Should we make `set_limit` do a `find_or_create_by` so it is always DB-backed and then drop this if?

File: app/views/users/edit_admin.html.erb [hunk 10]
```diff
@@ -107,6 +107,31 @@
...
+    <section>
+      <p>Admin transfer limit is currently: <strong><%= @limit ? number_to_currency(@limit.amount_cents / 100.0) : "None" %></strong></p>
+      <% if superadmin_signed_in? %>
+        <%= form_with url: update_admin_transfer_limit_user_path(@user), method: :post, data: { turbo_frame: "_top" } do |form| %>
+          <div class="field">
+            <%= form.label :limit_amount, "Set admin transfer limit (dollars)" %>
+            <%= form.number_field :limit_amount, step: 0.01, min: 0, value: @limit ? "%.2f" % (@limit.amount_cents / 100.0) : nil %>
+          </div>
+          <div class="actions flex">
+            <%= form.submit "Set admin transfer limit", data: { turbo_confirm: "Set the admin transfer limit for this user to $0.00?" }, onclick: "this.dataset.turboConfirm = 'Set the admin transfer limit for this user to $' +  this.form.limit_amount.value + '?'" %>
```
Comment (Luke-Oldenburg) [inline app/views/users/edit_admin.html.erb:122]: I think `$0.00` here might be wrong.

File: app/views/users/edit_admin.html.erb [hunk 11]
```diff
@@ -107,6 +107,31 @@
     </section>
   <% end %>
 
+  <% admin_tool "p-3" do %>
+    <h3 class="mb2 mt1">Admin Transfer Limit</h3>
```
Comment (Luke-Oldenburg) [inline app/views/users/edit_admin.html.erb]:
```suggestion
    <h3 class="mb2 mt1">Admin transfer limit</h3>
```

File: app/views/users/edit_admin.html.erb (hunk 10 above)
Comment (mahimtiwari) [reply app/views/users/edit_admin.html.erb:122]: turbo_confirm can't reference a dynamic ERB variable, so I am rewritting it using on click with the limit amount value in the input field

File: app/controllers/governance/admin/transfer/limits_controller.rb [hunk 12]
```diff
@@ -0,0 +1,48 @@
...
+        end
+
+        private
+
+        def set_user
+          @user = User.friendly.find(params[:id])
+        end
+
+        def set_limit
+          # Not using find_or_create here because on a brand new user if we do it would save the record to the db with val:null and then when we save the real user limit it would result in an error(subtraction of a nuber from nil) so we just create it in memeory and save it later when the limit is updated
```
Comment (Luke-Oldenburg) [inline app/controllers/governance/admin/transfer/limits_controller.rb]:
```suggestion
```

File: app/controllers/governance/admin/transfer/limits_controller.rb (hunk 9 above)
Comment (mahimtiwari) [reply app/controllers/governance/admin/transfer/limits_controller.rb]: I can't switch this line to find_or_create_by because it saves the record to the database the moment it's called, with amount_cents as nil. Then, when the actual limit is set and saved again, it tries to subtract a real number from that nil value, which throws an error. Keeping the original find_by || .new version avoids this — it only builds the record in memory, and the first time it's actually saved to the database is in update, once a real amount is already set.

**Review by Luke-Oldenburg — APPROVED** [review]
lgtm
